### DISCUSSION TITLE
Godot scenes - Sonic worlds next treats them differently than I'm used to, can someone explain?

### DISCUSSION THREAD

**Discussion by Cruisemang** [Techokami/SonicWorldsNext#174] [General]
Hi,
I have made some really basic 'games' in Godot, and the way I've always set up the player and every interactable object is to save them as their own scenes, that have their own scripts etc.
Eg: Sonic and Tails even though very similar would be saved as their own scenes with near duplicate scripts attached to them (but tails would have flying functionality instead) and in their scenes their sprite and hitboxes would be different.
Other objects like a monitor or a spring would also be their own scene.

What I'm getting at is Sonic Worlds Next is obviously not made by a novice like myself and they know what they're doing. But the way it's done is just so hard for a novice like me to understand. Could someone explain their method to me? It's not the scene method as I've written above, unless I'm missing something?

I've made very straight forward mario-type platformers, but trying to go next level and doing a sonic style platformer is so hard and I've had a hard time on youtube and havent really found anything.

Any help would be amazing - 
thank you.

**Comment by Renhoex** [comment]
For worlds next we used a variable to detect what character the player is using currently, while the multiple scenes method is easy to understand, keeping it in sync with the main node would require a tonne of time to set up and updating one characters would mean updating the rest as well.

you can take a look at the character select script for how it sets the characters up
`Global.PlayerChar1 = Global.CHARACTERS.SONIC`

for player actions, the air script is a good one to look at,
to determine what ability to use there's a match case function that loops through the different players and runs those scripts if it finds a match
`match (parent.character):
					parent.CHARACTERS.SONIC:`

for setting up the initial set up for the characters, look in the ready function of the player script around line 340,
if it finds a match for the characters it assigns the hitbox and sets the character sprites and other settings. (like removing sonics sprites for tails)

There's an enum for the characters around line 66 that has a list of the different characters, tho if you update that you should check in the Global script for the same enum (in retrospect I should have made the player script just refer to the global script instead)

**Reply by Cruisemang** [reply]
I think it's a bit beyond me right now, hopefully I can come back and have a look when I've played with Godot a bit more. It's funny that at first glance, making a mario game then a sonic game doesn't seem that far apart until you try and make it yourself :/ Thanks for your reply btw.

**Comment by Candescence** [comment]
Explaining how this all works and why it works the way it does is probably one thing that should be added to the documentation/wiki. Said wiki hasn't been properly updated in a _while_ now, and quite a few things in Worlds NEXT is a bit different to how many people learning Godot would be accustomed to.

Edit: Also, side-note, the way the player system works doesn't feel very "scalable". As it is now, it looks like it'd be a pain in the rear to implement more than two players/bots.

**Reply by Cruisemang** [reply]
I think it's a bit beyond my capabilities right now, but I appreciate the reply!

**Reply by Candescence** [reply]
I was frankly responding more to Ren and trying to address the World NEXT team in general, but no worries.
